static/app/components/forms/teamSelector.tsx

```diff
@@ -0,0 +1,265 @@
+import withApi from 'app/utils/withApi';
+import withOrganization from 'app/utils/withOrganization';
+import withTeams from 'app/utils/withTeams';
+
+import IdBadge from '../idBadge';
+import Tooltip from '../tooltip';
+
+const UnassignedWrapper = styled('div')`
+  display: flex;
+  align-items: center;
+`;
+
+const StyledIconUser = styled(IconUser)`
+  margin-left: ${space(0.25)};
+  margin-right: ${space(1)};
+  color: ${p => p.theme.gray400};
+`;
+
+// An option to be unassigned on the team dropdown
+const unassignedOption = {
+  value: null,
+  label: (
+    <UnassignedWrapper>
+      <StyledIconUser size="20px" />
+      {t('Unassigned')}
+    </UnassignedWrapper>
+  ),
+  searchKey: 'unassigned',
+  actor: null,
+  disabled: false,
+};
+
+// Ensures that the svg icon is white when selected
+const unassignedSelectStyles = {
+  option: (provided, state: any) => ({
+    ...provided,
+    svg: {
+      color: state.isSelected && state.theme.white,
+    },
+  }),
+};
+
+type Props = {
+  api: Client;
+  organization: Organization;
+  teams: Team[];
+  onChange: (value: any) => any;
+  /**
+   * Function to control whether a team should be shown in the dropdown
+   */
+  teamFilter?: (team: Team) => boolean;
+  /**
+   * Can be used to restrict teams to a certain project and allow for new teams to be add to that project
+   */
+  project?: Project;
+  /**
+   * Controls whether the value in the dropdown is a team id or team slug
+   */
+  useId?: boolean;
+  includeUnassigned?: boolean;
+} & ControlProps;
+
+type TeamActor = {
+  type: 'team';
+  id: string;
+  name: string;
+};
+
+type TeamOption = {
+  value: string | null;
+  label: React.ReactElement;
+  searchKey: string;
+  actor: TeamActor | null;
+  disabled?: boolean;
+};
+
+type State = {
+  options: TeamOption[];
+};
+
+class TeamSelector extends React.Component<Props, State> {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      options: this.teamOptions,
+    };
+  }
+
+  // TODO(ts) This type could be improved when react-select types are better.
+  selectRef = React.createRef<any>();
+
+  get teamOptions() {
+    const {teams, teamFilter, includeUnassigned, project} = this.props;
+    const filteredTeams = teamFilter ? teams.filter(teamFilter) : teams;
+    if (project) {
+      const teamsInProjectIdSet = new Set(project.teams.map(team => team.id));
+      const teamsInProject = filteredTeams.filter(team =>
+        teamsInProjectIdSet.has(team.id)
+      );
+      const teamsNotInProject = filteredTeams.filter(
+        team => !teamsInProjectIdSet.has(team.id)
+      );
+
+      return [
+        ...teamsInProject.map(this.createTeamOption),
+        ...teamsNotInProject.map(this.createTeamOutsideProjectOption),
+        ...(includeUnassigned ? [unassignedOption] : []),
+      ];
+    }
+    return [
+      ...filteredTeams.map(this.createTeamOption),
+      ...(includeUnassigned ? [unassignedOption] : []),
+    ];
+  }
```

Comment:
So, this is a more general comment. But I might encourage you to use more blank line in your code to separate out 'blocks' or sections of logic.

For example, I might write this more like this.

```tsx
    const {teams, teamFilter, includeUnassigned, project} = this.props;
    const filteredTeams = teamFilter ? teams.filter(teamFilter) : teams;

    if (project) {
      const teamsInProjectIdSet = new Set(project.teams.map(team => team.id));
      const teamsInProject = filteredTeams.filter(team =>
        teamsInProjectIdSet.has(team.id)
      );
      const teamsNotInProject = filteredTeams.filter(
        team => !teamsInProjectIdSet.has(team.id)
      );

      return [
        ...teamsInProject.map(this.createTeamOption),
        ...teamsNotInProject.map(this.createTeamOutsideProjectOption),
        ...(includeUnassigned ? [unassignedOption] : []),
      ];
    }

    return [
      ...filteredTeams.map(this.createTeamOption),
      ...(includeUnassigned ? [unassignedOption] : []),
    ];
```

As a general rule, I put whitespace around my control flow statements (ifs, fors, etc).

I have to think about some other rules. But I think there is a sense of 'giving code room to breath' which makes it more readable.

What do you think?

Reply:
dang yea this makes a lot of sense and feels much nicer. tyty for the tips